File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<int> next(2, 0), cur(2, 0);
        
        for(int i = n-1; i >= 0; i--) {
            cur[0] = max(prices[i]+next[1], next[0]);
            cur[1] = max(-prices[i]+next[0], next[1]);
            
            next = cur;
        }
        
        return next[1];
    }
};

File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.js
/**
 * @param {number[]} prices
 * @return {number}
 */

var maxProfit = function(prices) {
    const n = prices.length;
    let dp = new Array(n+1);
    for(let i = 0; i <= n; i++) {
        dp[i] = new Array(2);
        dp[i][0] = dp[i][1] = 0;
    } 

    for(let i = n-1; i >= 0; i--) {
        dp[i][0] = Math.max(dp[i+1][1]+prices[i], dp[i+1][0]);
        dp[i][1] = Math.max(dp[i+1][0]-prices[i], dp[i+1][1]);
    }
    
    return dp[0][1]
};